drivers/sensor/nordic/qdec_nrfx/qdec_nrfx.c

```diff
@@ -173,81 +174,73 @@ static const struct sensor_driver_api qdec_nrfx_driver_api = {
 	.trigger_set  = qdec_nrfx_trigger_set,
 };
 
-#ifdef CONFIG_PM_DEVICE
-static int qdec_nrfx_pm_action(const struct device *dev,
-			       enum pm_device_action action)
+static int qdec_pm_suspend(const struct device *dev)
 {
 	const struct qdec_nrfx_config *config = dev->config;
-	int ret = 0;
+
+	nrfx_qdec_disable(&config->qdec);
+	qdec_nrfx_gpio_ctrl(dev, false);
+
+	return pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
+}
+
+static int qdec_pm_resume(const struct device *dev)
+{
+	const struct qdec_nrfx_config *config = dev->config;
+	int ret;
+
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	qdec_nrfx_gpio_ctrl(dev, true);
+	nrfx_qdec_enable(&config->qdec);
+
+	return 0;
+}
+
+static int qdec_nrfx_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	int ret = -ENOTSUP;
 
 	switch (action) {
 	case PM_DEVICE_ACTION_RESUME:
-		ret = pinctrl_apply_state(config->pcfg,
-					  PINCTRL_STATE_DEFAULT);
-		if (ret < 0) {
-			return ret;
-		}
-		qdec_nrfx_gpio_ctrl(dev, true);
-		nrfx_qdec_enable(&config->qdec);
-		break;
-
-	case PM_DEVICE_ACTION_TURN_OFF:
-		/* device must be uninitialized */
-		nrfx_qdec_uninit(&config->qdec);
-		ret = pinctrl_apply_state(config->pcfg,
-					  PINCTRL_STATE_SLEEP);
-		if (ret < 0) {
-			return ret;
-		}
+		ret = qdec_pm_resume(dev);
 		break;
 
 	case PM_DEVICE_ACTION_SUSPEND:
 		/* device must be suspended */
-		nrfx_qdec_disable(&config->qdec);
-		qdec_nrfx_gpio_ctrl(dev, false);
-		ret = pinctrl_apply_state(config->pcfg,
-					  PINCTRL_STATE_SLEEP);
-		if (ret < 0) {
-			return ret;
-		}
+		ret = qdec_pm_suspend(dev, true);
 		break;
 	default:
-		return -ENOTSUP;
+		break;
 	}
 
 	return ret;
 }
-#endif /* CONFIG_PM_DEVICE */
 
 static int qdec_nrfx_init(const struct device *dev)
 {
-	const struct qdec_nrfx_config *dev_config = dev->config;
-
-	dev_config->irq_connect();
+	const struct qdec_nrfx_config *config = dev->config;
+	int ret;
+	nrfx_err_t nerr;
 
-	int err = pinctrl_apply_state(dev_config->pcfg, PINCTRL_STATE_DEFAULT);
+	config->irq_connect();
 
-	if (err < 0) {
-		return err;
+	nerr = nrfx_qdec_init(&config->qdec, &config->config, qdec_nrfx_event_handler, (void *)dev);
+	if (nerr != NRFX_SUCCESS) {
+		return (nerr == NRFX_ERROR_INVALID_STATE) ? -EBUSY : -EFAULT;
 	}
 
-	nrfx_err_t nerr = nrfx_qdec_init(&dev_config->qdec,
-					 &dev_config->config,
-					 qdec_nrfx_event_handler,
-					 (void *)dev);
-
-	if (nerr == NRFX_ERROR_INVALID_STATE) {
-		LOG_ERR("qdec already in use");
-		return -EBUSY;
-	} else if (nerr != NRFX_SUCCESS) {
-		LOG_ERR("failed to initialize qdec");
-		return -EFAULT;
+	if (IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)) {
+		ret = qdec_pm_suspend(dev, false);
+		if (ret < 0) {
+			return ret;
+		}
```

Comment:
This should not be needed, `pm_device_driver_init(dev, qdec_nrfx_pm_action);` does it automatically if the zephyr,pm-device-runtime-auto; flag is set for the node in the devicetree.

Reply:
It only sets flag but it is expected that after turning on device is in the suspend state. https://docs.zephyrproject.org/latest/services/pm/device.html#device-power-management-states

Reply:
The flag is used by the call to `pm_device_driver_init()` :)

```
int pm_device_driver_init(const struct device *dev,
			  pm_device_action_cb_t action_cb)
{
         ...

	/* If device will have PM device runtime enabled */
	if (IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME) &&
	    atomic_test_bit(&pm->flags, PM_DEVICE_FLAG_RUNTIME_AUTO)) {
		/* Init into suspend mode.
		 * This saves a SUSPENDED->ACTIVE->SUSPENDED cycle.
		 */
		pm_device_init_suspended(dev);
		return 0;
	}
```

so it will indeed be suspended automatically on init

Reply:
not sure I understood but `pm_device_init_suspended(dev);` is only setting state to SUSPENDED and do not call any action. I was clarifying it with @JordanYates https://discord.com/channels/720317445772017664/733037675786797127/1286738310085214350 and from what i understood init should leave the device in the suspended state.

Reply:
I see, I did think it called `SUSPEND` explicitly, then yes, if we need to call it on boot (which maybe we don't? should it not be in its reset state already?)